src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -30,17 +37,38 @@ private void InitSort(CultureInfo culture)
             }
             else
             {
-                Interop.Globalization.ResultCode resultCode = Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(_sortName), out _sortHandle);
-                if (resultCode != Interop.Globalization.ResultCode.Success)
-                {
-                    _sortHandle.Dispose();
+                _isAsciiEqualityOrdinal = (_sortName == "en-US" || _sortName == "");
 
-                    if (resultCode == Interop.Globalization.ResultCode.OutOfMemory)
-                        throw new OutOfMemoryException();
+                lock (_lock)
+                {
+                    if (s_sortNameToSortHandleCache == null)
+                    {
+                        s_sortNameToSortHandleCache = new Dictionary<string, IntPtr>();
+                    }
 
-                    throw new ExternalException(SR.Arg_ExternalException);
+                    if (!s_sortNameToSortHandleCache.TryGetValue(_sortName, out _sortHandle))
+                    {
+                        s_sortNameToSortHandleCache.Add(_sortName, (_sortHandle = GetSortHandle(_sortName)));
+                    }
                 }
-                _isAsciiEqualityOrdinal = (_sortName == "en-US" || _sortName == "");
+            }
+        }
+
+        private static IntPtr GetSortHandle(string sortName)
+        {
+            var resultCode = Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(sortName), out IntPtr sortHandle);
+            if (resultCode == Interop.Globalization.ResultCode.Success)
+            {
+                return sortHandle;
+            }
+            else
+            {
+                Interop.Globalization.CloseSortHandle(sortHandle);
```

Comment:
This is not correct. GetSortHandle does not return anything on error, and trying to call CloseSortHandle on NULL here will just crash in the native code.

Reply:
good catch, I have removed it